### PR TITLE
Switch to new labelling scheme for our GitHub runners

### PR DESCRIPTION
**Pull request by potiuk**
---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

**Comment by potiuk** [comment]
I also improved significantly the output of selective checks tests when they see something unexpected:

It was rather cryptic with the regular assertion so I added a bit more custom/coloured output and more accurate and explicit description showing what's expected, what's received and what was the output received (as nicely formatted dict).

Screenshot from 2023-10-22 17-38-14

**Comment by potiuk** [comment]
But I need to make it works:

```
Requested labels: "\"[\\\"self-hosted\\\", \\\"airflow-runner\\\", \\\"vm-runner\\\", \\\"X64\\\"]\""
```

File: dev/breeze/src/airflow_breeze/global_constants.py
Comment (hussein-awala) [inline dev/breeze/src/airflow_breeze/global_constants.py]: I don't seed the labels attached to [the runners](https://github.com/apache/airflow/actions/runners?tab=self-hosted); we need to check before merging to avoid blocking the committers' PRs.

Comment (hussein-awala) [reply dev/breeze/src/airflow_breeze/global_constants.py]: I think this should passed as an array instead of a string; it will be serialized to JSON by GHA :thinking:

Comment (potiuk) [reply dev/breeze/src/airflow_breeze/global_constants.py]: > I don't seed the labels attached to [the runners](https://github.com/apache/airflow/actions/runners?tab=self-hosted); we need to check before merging to avoid blocking the committers' PRs.

Yeah. This one won't get green until that works, so no worries :)

> I think this should passed as an array instead of a string; it will be serialized to JSON by GHA 🤔

It's complicated. We can only print strings as outputs and when needed they need to be converted to arrays with "toJSON" method - but its not obvious how so I am playing with it now to get the right  combination

:shrug: GA woes.

Comment (potiuk) [reply dev/breeze/src/airflow_breeze/global_constants.py]: Or rather fromJSON :facepalm:

Comment (hussein-awala) [reply dev/breeze/src/airflow_breeze/global_constants.py]: You can try:
```
runs-on: ${{ toJSON(steps.selective-checks.outputs.runs-on) }}
```
in the list of outputs

Comment (hussein-awala) [reply dev/breeze/src/airflow_breeze/global_constants.py]: your last commit works :tada: 
```
Requested labels: self-hosted, airflow-runner, vm-runner, X64
```

Comment (potiuk) [reply dev/breeze/src/airflow_breeze/global_constants.py]: > runs-on: ${{ toJSON(steps.selective-checks.outputs.runs-on) }}

Unfortunately not.... Outputs are always Unicode strings - so converting them to JSON just causes subsequent encoding :shrug: ... We need to run `fromJSON` in all the places where the outputs are used as part of workflow definition ...


I think I got it right now... But indeed our runners do not have the right labels defined and do not pick the jobs up:


> Requested labels: self-hosted, airflow-runner, vm-runner, X64
Job defined at: apache/airflow/.github/workflows/ci.yml@refs/pull/35109/merge
Waiting for a runner to pick up this job...

**Review by eladkal — APPROVED** [review]

**Review by hussein-awala — APPROVED** [review]

**Comment by potiuk** [comment]
There is a small risk that this one will fail in `build-image` workflow after I merge it in main so keep fingers crossed

**Pull request merged**
